### PR TITLE
Refactored parser code by removing any out of state actions

### PR DESCRIPTION
**Pull request by ShreyasRmsft**
Also added common negative tests for jasmine

### CODE REVIEW

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/MochaTestResultParser.cs [hunk 1]
```diff
@@ -153,6 +123,8 @@ private bool AttemptMatch(ITestResultParserState state, LogData logData)
                 }
             }
 
+            state.NoPatternMatched(logData.Line, stateContext);
```
Comment (singhsarab) [inline Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/MochaTestResultParser.cs]: Nit: PeformNoPatternMatchedAction()

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Common/NodeParserStateBase.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            this.logger = logger;
+            this.parserName = parserName;
+            this.stateName = this.GetType().Name;
+            this.telemetryDataCollector = telemetryDataCollector;
+            this.attemptPublishAndResetParser = parserResetAndAttempPublish;
+        }
+
+        /// <summary>
+        /// Default implemenation that checks for the constraint for the next expected match
+        /// If the number of lines within whic the next match expected falls to 0 this resets the parser
```
Comment (singhsarab) [inline Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Common/NodeParserStateBase.cs]: >whic [](start = 42, length = 4)

Nit:which

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingStackTraces.cs [hunk 3]
```diff
@@ -179,9 +180,34 @@ private Enum PassedTestsSummaryMatched(Match match, AbstractParserStateContext s
...
             return MochaParserStates.ExpectingTestRunSummary;
         }
+
+        /// <summary>
+        /// If none of the patterns matched then considers adding the current line to stack trace
+        /// based on whether a stack trace start has been encountered
+        /// </summary>
+        /// <param name="line">Current line</param>
+        /// <param name="stateContext">State context object containing information of the parser's state</param>
+        /// <returns></returns>
```
Comment (singhsarab) [inline Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingStackTraces.cs]: Add the return.

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingTestRunSummary.cs [hunk 4]
```diff
@@ -164,7 +165,7 @@ private Enum FailedTestsSummaryMatched(Match match, AbstractParserStateContext s
             mochaStateContext.StackTracesToExpectPostSummary = totalFailed;
 
             // Do we want transition logs here?
-            this.logger.Info($"MochaTestResultParser : ExpectingTestRunSummary : Transitioned to state ExpectingStackTraces" +
+            this.logger.Info($"{this.parserName} : {this.stateName} : Transitioned to state ExpectingStackTraces" +
                 $" at line {mochaStateContext.CurrentLineNumber}.");
             return MochaParserStates.ExpectingStackTraces;
```
Comment (singhsarab) [inline Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingTestRunSummary.cs]: >return MochaParserStates.ExpectingStackTraces; [](start = 12, length = 46)

Need to have a check, expectedlines and add tests.

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/MochaTestResultParser.cs (hunk 1 above)
Comment (ShreyasRmsft) [reply Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/MochaTestResultParser.cs]: done

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingStackTraces.cs (hunk 3 above)
Comment (ShreyasRmsft) [reply Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingStackTraces.cs]: done

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Common/NodeParserStateBase.cs (hunk 2 above)
Comment (ShreyasRmsft) [reply Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Common/NodeParserStateBase.cs]: done

File: Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingTestRunSummary.cs (hunk 4 above)
Comment (ShreyasRmsft) [reply Agent.Plugins.Log.TestResultParser.Parser/Parser/Node/Mocha/States/MochaParserStateExpectingTestRunSummary.cs]: done

**Pull request merged**